common/reflect_resource.go

```diff
@@ -74,6 +218,111 @@ func MustSchemaPath(s map[string]*schema.Schema, path ...string) *schema.Schema
 	return sch
 }
 
+func CustomizeSchemaPath(s map[string]*schema.Schema, path ...string) *CustomizableSchema {
+	sch := MustSchemaPath(s, path...)
+	return &CustomizableSchema{Schema: sch}
+}
+
+type CustomizableSchema struct {
+	Schema *schema.Schema
+}
+
+func (s *CustomizableSchema) SetOptional() *CustomizableSchema {
+	s.Schema.Optional = true
+	s.Schema.Required = false
+	return s
+}
+
+func (s *CustomizableSchema) SetComputed() *CustomizableSchema {
+	s.Schema.Computed = true
+	return s
+}
+
+func (s *CustomizableSchema) SetDefault(value any) *CustomizableSchema {
+	s.Schema.Default = value
+	s.Schema.Optional = true
+	s.Schema.Required = false
+	return s
+}
+
+// SetReadOnly sets the schema to be read-only (i.e. computed, non-optional).
+// This should be used for fields that are not user-configurable but are returned
+// by the platform.
+func (s *CustomizableSchema) SetReadOnly() *CustomizableSchema {
+	s.Schema.Optional = false
+	s.Schema.Required = false
+	s.Schema.MaxItems = 0
```

Comment:
Yeah good question, I was just trying to stay consistent with the function introduced by Miles [here](https://github.com/databricks/terraform-provider-databricks/pull/3044/files#diff-50f3dff318c929631c545c2a0286e77407a68f99cc0d99949e6151f1a4f45d80R104-R109)

Reply:
@mgyucht Any idea about this?

Reply:
One more thing is that this won't work correctly if we call `SetReadOnly` after `SetMaxItems` as then the value would be overwritten. We would also need a check to prevent this. 

```
func (s *CustomizableSchema) SetMaxItems(value int) *CustomizableSchema {
	s.Schema.MaxItems = value
	return s
}
```

Reply:
Yeah synced with Miles, the idea is that when a field is `Computed` or read-only, terraform doesn't really want us to set any constraint to the value itself (max items is an example).

Regarding calling `SetMaxItems` after calling `SetReadOnly`, it should probably not be a problem because people should see the schema being undesirable in the diff schema check or just see some failures.